File: CS/WindowsFormsApplication72/SearchHelper.cs
using DevExpress.XtraEditors;
using DevExpress.XtraNavBar;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApplication72
{
    public enum SearchCriteria
    {
        Contains,
        StartsWith,
        Equals
    }

    public static class SearchHelper
    {
        static NavBarControl navBarControl;
        static TextEdit textEdit;
        static SearchCriteria criteria;

        public static void CreateSearchPanel(NavBarControl _navBarControl, SearchCriteria _criteria)
        {
            if (_navBarControl != null)
            {
                navBarControl = _navBarControl;
                criteria = _criteria;
                textEdit = new TextEdit();
                NavBarGroup navBarGroup = new NavBarGroup();
                _navBarControl.Groups.Insert(0, navBarGroup);
                navBarGroup.GroupStyle = NavBarGroupStyle.ControlContainer;
                navBarGroup.Caption = "Search";
                
                navBarGroup.ControlContainer.Controls.Add(textEdit);
                textEdit.Dock = System.Windows.Forms.DockStyle.Fill;
                textEdit.Visible = true;
                navBarGroup.GroupClientHeight = 26;
                navBarGroup.Expanded = true;
                _navBarControl.Groups.Add(navBarGroup);

                _navBarControl.CustomDrawGroupCaption += navBarControl_CustomDrawGroupCaption;
                textEdit.EditValueChanged += textEdit_EditValueChanged;
            }
        }

        static void navBarControl_CustomDrawGroupCaption(object sender, DevExpress.XtraNavBar.ViewInfo.CustomDrawNavBarElementEventArgs e)
        {
            if (e.Caption == "Search")
            {
                Rectangle rect = e.RealBounds;
                rect.Inflate(-10, -5);

                e.Graphics.DrawString(e.Caption, e.Appearance.Font, Brushes.Black, rect);
                e.Handled = true;
            }
        }

        static void textEdit_EditValueChanged(object sender, EventArgs e)
        {
            navBarControl.BeginUpdate();
            foreach (NavBarGroup group in navBarControl.Groups)
            {
                foreach (NavBarItemLink link in group.ItemLinks)
                {
                    switch (criteria){
                        case SearchCriteria.Contains:
                            link.Visible = link.Caption.Contains(textEdit.Text);
                            break;
                        case SearchCriteria.StartsWith:
                             link.Visible = link.Caption.StartsWith(textEdit.Text);
                            break;
                        case SearchCriteria.Equals:
                            link.Visible = link.Caption.Equals(textEdit.Text);
                            break;
                    }
                    if (group.VisibleItemLinks.Count == 0)
                        group.Visible = false;
                    else
                        group.Visible = true;
                }
            }
            navBarControl.EndUpdate();
        }

    }
}

File: VB/WindowsFormsApplication72/SearchHelper.vb
Imports DevExpress.XtraEditors
Imports DevExpress.XtraNavBar
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace WindowsFormsApplication72
	Public Enum SearchCriteria
		Contains
		StartsWith
		Equals
	End Enum

	Public NotInheritable Class SearchHelper

		Private Sub New()
		End Sub

		Private Shared navBarControl As NavBarControl
		Private Shared textEdit As TextEdit
		Private Shared criteria As SearchCriteria

		Public Shared Sub CreateSearchPanel(ByVal _navBarControl As NavBarControl, ByVal _criteria As SearchCriteria)
			If _navBarControl IsNot Nothing Then
				navBarControl = _navBarControl
				criteria = _criteria
				textEdit = New TextEdit()
				Dim navBarGroup As New NavBarGroup()
				_navBarControl.Groups.Insert(0, navBarGroup)
				navBarGroup.GroupStyle = NavBarGroupStyle.ControlContainer
				navBarGroup.Caption = "Search"

				navBarGroup.ControlContainer.Controls.Add(textEdit)
				textEdit.Dock = System.Windows.Forms.DockStyle.Fill
				textEdit.Visible = True
				navBarGroup.GroupClientHeight = 26
				navBarGroup.Expanded = True
				_navBarControl.Groups.Add(navBarGroup)

				AddHandler _navBarControl.CustomDrawGroupCaption, AddressOf navBarControl_CustomDrawGroupCaption
				AddHandler textEdit.EditValueChanged, AddressOf textEdit_EditValueChanged
			End If
		End Sub

		Private Shared Sub navBarControl_CustomDrawGroupCaption(ByVal sender As Object, ByVal e As DevExpress.XtraNavBar.ViewInfo.CustomDrawNavBarElementEventArgs)
			If e.Caption = "Search" Then
				Dim rect As Rectangle = e.RealBounds
				rect.Inflate(-10, -5)

				e.Graphics.DrawString(e.Caption, e.Appearance.Font, Brushes.Black, rect)
				e.Handled = True
			End If
		End Sub

		Private Shared Sub textEdit_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
			navBarControl.BeginUpdate()
			For Each group As NavBarGroup In navBarControl.Groups
				For Each link As NavBarItemLink In group.ItemLinks
					Select Case criteria
						Case SearchCriteria.Contains
							link.Visible = link.Caption.Contains(textEdit.Text)
						Case SearchCriteria.StartsWith
							 link.Visible = link.Caption.StartsWith(textEdit.Text)
						Case SearchCriteria.Equals
							link.Visible = link.Caption.Equals(textEdit.Text)
					End Select
					If group.VisibleItemLinks.Count = 0 Then
						group.Visible = False
					Else
						group.Visible = True
					End If
				Next link
			Next group
			navBarControl.EndUpdate()
		End Sub

	End Class
End Namespace
